### ISSUE TITLE
json-api run with query store returns inconsistent data, when multi-party is querying at the same time

### ISSUE THREAD

**Issue by lima-da** [digital-asset/daml#4105]
Here is the scenario describes the issue:

When `partyA` and `partyB` both can exercise a nonconsuming choice on contractC to create multiple contracts (e.g. spamming contractC.A1, contractC.A2...contractC.B1, contractC.B2), if `partyA` and `partyB` queries `/contracts/search` immediately right after their choice exercise interchangeably/concurrently, the query store gets confused on which cached query result to serve to which party. 

As a result, from `partyA` 's perspective, the query result of `/contracts/search POST` for specific `templateId` has less data, compared to the result of a post-filter of `/contracts/search GET` by using that `templateId `. 

After run `truncate table template_id cascade;` directly on the query store database. and do a `/contracts/search POST` again as `partyA`, the result for that specific `templateId` matches the result of `/contracts/search GET` however, if partyB exercise&query again, the result becomes indeterministic again

Please see the below model with scenario/steps for reproducing. and reach out to @lima-da if you cannot reproduce.

```
-- Copyright (c) 2020 The DAML Authors. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

daml 1.2
module Main where

import DA.List


template Chat
  with
    operator : Party
    creator : Party
    members : [Party]
    name : Text
  where
    signatory operator
    observer members

    choice Chat_AddMembers : ContractId Chat
      with
        member : Party
        newMembers : [Party]
      controller member
      do
        create this with members = dedup $ newMembers ++ members

    nonconsuming choice Chat_PostMessage : ContractId Message
      with
        poster : Party
        message : Text
      controller poster
      do
        create Message with sender = poster, recipients = members, ..


template Message
  with
    sender : Party
    message : Text
    recipients : [Party]
  where
    signatory sender
    observer recipients


setup = scenario do
  operator <- getParty "Operator"
  alice <- getParty "Alice"
  bob <- getParty "Bob"

  -- operator creates a chat on behalf of Alice
  chatCid <- operator `submit` create Chat with creator = alice, members = [alice], name = "sleek-cheek", ..

  -- alice sends a message
  helloCid <- alice `submit` exercise chatCid Chat_PostMessage with poster = alice, message = "first message without invite Bob, and query as Alice after"

  -- do a /contracts/search POST with Alice's party token
  -- {"%templates":[{"moduleName":"Main","entityName":"Chat"},{"moduleName":"Main","entityName":"Message"}]}
  -- everything from Alice match /contracts/search GET

  -- operator(alice) adds bob(as a member) to chat. Chat gets recreated
  chatCid <- operator `submit` exercise chatCid Chat_AddMembers with member = operator, newMembers = [bob]

  -- act as Bob, and bob sends a message
  goodbyeCid <- bob `submit` exercise chatCid Chat_PostMessage with poster = bob, message = "Bob enters, and post first message, query as Bob after"

  -- repeat /contracts/search POST but with Bob's party token
  -- then /contracts/search POST with Alice's party token, and you can see there are missing results compared with /contracts/search GET
  -- results matches after you run "truncate table template_id cascade;"

  return ()
```

**Comment by S11001001** [comment]
There are a lot of things described in your post that may not actually be issues, depending on what you mean, and certainly a few that may be issues, so I'll focus on some clarifying questions:

Let _earlier_ mean "reflective of the ledger state at a prior end offset", _later_ mean "reflective of the ledger state at a later end offset", and _same time_ mean the remaining possibility. Considering the last two lines,

1. Suppose you make the POST as Bob. Let that be result A.
2. Suppose you make the POST as Alice at the same time as the previous. Let that result be B.
    - A may be _earlier_, _later_, or _same time_ as B; there is no reason to expect any particular one.
3. Then you _subsequently_, after result A is produced (i.e. the POST response is complete), make the POST as Alice. Let that result be C.
    - If C is _earlier_ than A, that is a problem.
4. Then, _subsequently_, truncate as described in your post, without making any additional ledger changes. (You've described what happens if you exercise again after truncate; I want to be clear that this is _not_ what I mean.)
5. Making the POST once more as Alice, let that result be D. (Will use this below.)

The description "immediately right after their choice exercise interchangeably/concurrently" seems to imply producing results A and B; if that is the case then as I mentioned there are no reasonable assumptions to make about their relative time. However if what you mean is results A and C, then we definitely have something to explore.

It is not worth comparing to search GET due to #3254 . It is _especially_ not worth comparing its concurrent behavior, even if #3254 was fixed, because it spends longer in real-time talking to the ledger.

----

> After run truncate table template_id cascade; directly on the query store database. and do a /contracts/search POST again as partyA, the result for that specific templateId matches the result of /contracts/search GET 

It's difficult to draw any conclusions from this, because after all, it is reasonable to expect that this POST will produce a _later_ or _same time_ result than the earlier POST.

It is also difficult to draw any conclusions about not observing the reported behavior, or observing it less often, under concurrent conditions following the truncate you describe, because this action causes the JSON API to spend more time talking to the ledger as with GET above, as with GET, implying that it is likelier to reach end-of-ledger at a later real time. (Incidentally we are considering a change to database search related to #3881 that will cause JSON API to stop at earlier ledger ends, thereby causing the truncate "workaround" as you describe to stop behaving in the way you describe.)

However, if results C and D are not _same time_, then that is a problem worth exploring.

----

As described, there is no reason to expect A and B to be _same time_, and therefore talking about "consistency" with respect to them is difficult.

However, _if a single query returns results that are reflective of an impossible ledger state, that is definitely a problem_. For example, if the results of a single query produce a contract, and also a contract produced by the same exercise that archived that contract, that is a problem. Did you observe this at any time in your experiments?

----

To summarize:

* [x] There is no guarantee about the relative time of A and B.
* [ ] Is C _earlier_ than A? If so, that is a problem.
* [ ] Is C _same time_ as D? If not, that is a problem.
* [ ] Do any query results reflect, by themselves, an impossible ledger state? If so, that is a problem.

**Comment by lima-da** [comment]
Sorry about the confusion here, we spotted the problem in the DABL chat app, that's why I describe/assume the problem happens when exercise choice on behalf of two parties interchangeably/concurrently. That may imply the scenario of producing results A and B. However, I meant the problem is in order of results A and C.

Let me rephrase here:

The first thing caught our attention was the result from `POST contract/search` is less than `GET contract/search` for the same `templateId` contract for one party. And by shooting in the dark, I tried truncating the query store DB for json-api. The mismatch goes away for that party. Hence my guess is the problem may be tied to the query store.

To narrow the problem further, I tried to reproduce this with a toy model using sandbox/navigator/json-api. As described in the `daml scenario` comment above, the steps are sequentially performed as Alice and Bob one after another, and manually observing the result from navigator and json-api response before taking the next steps. I can constantly reproduce the problem.

"Is C earlier than A ?", not sure what "earlier" means here. in our case, C is produced after A, but C has less data and missing "older" data.

Let me know if you have any other questions.

**Comment by lima-da** [comment]
@S11001001 This issue was spotted on SDK version 0.13.40,  it seems no longer reproducible after we upgraded SDK to 0.13.46.

**Comment by S11001001** [comment]
The way we limit query has significantly changed, and remaining issues with GET due to #3254.
